keps/sig-node/4580-deprecate-kubelet-runonce/README.md

```diff
@@ -0,0 +1,354 @@
+```
+
+### Introduction LegacyNodeRunOnceMode feature gate
+
+With the introduction of the `LegacyNodeRunOnceMode` feature gate, Kubernetes aims to guide users through the deprecated Runonce mode. Unless this feature gate is enabled, kubelet will refuse to start when the `--runonce` command line flag is set.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- N/A
+
+##### Integration tests
+
+- N/A
+
+##### e2e tests
+
+- N/A
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature gate `LegacyNodeRunOnceMode` is introduced, is enable by default. Disable this feature gate will fail the kubelet on startup with Runonce mode enable.
+- Mark the `RunOnce` field in `KubeletConfiguration` as deprecated.
+
+#### Beta
+
+- `LegacyNodeRunOnceMode` feature gate is disable by default.
+- Failed when starting kubelet in Runonce mode.
+
+#### GA
+
+- We make the `LegacyNodeRunOnceMode` feature gate disable by default and cannot be enable.
+- Comment the `RunOnce` field in KubeletConfiguration as 'no longer has any effect', and remove the kubelet's `runonce` flag.
+- Remove kubelet Runonce mode.
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+- N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: LegacyNodeRunOnceMode
+  - Components depending on the feature gate: kubelet
+  - Will enabling / disabling the feature require downtime of the control
+    plane? Yes. Flag must be set on kubelet start. To disable, kubelet must be restarted. Hence, there would be brief control component downtime on a given node.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? Yes. See above; disabling would require brief node downtime.
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. Using the feature gate is the only way to enable/disable this feature.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+N/A
+
+###### Are there any tests for feature enablement/disablement?
+
+N/A
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+In the alpha `stage`, this feature is enable by default.
+
+Cluster operators can test the behavior by enabling the feature gate.
+
+In the beta `stage`, this feature is enable by default. With this feature disable, the kubelet will refuse to start if is still using Runonce mode.
+
+Cluster operators can reinstate the mode by explicitly enabling the feature gate.
```

Comment:
Is there a metric or a way for a cluster administrator to verify if this functionality is being used other than checking the kubelet config file and invocation?

Reply:
No, there is no other way for cluster administrator to verify whether this feature is enabled.

Reply:
We _could_ have kubelets in RunOnce mode emit a special (alpha?) metric, and delay the switch to off-by-default until all supported releases do emit that metric. If we want.

Reply:
Considering that this KEP is to deprecate a feature, I am not sure if we have enough motivation to add a new metric for it. Do we have similar implementations in the past?

Reply:
No, you're right we can get away without adding one.